skyportal/handlers/api/internal/annotations_info.py

```diff
@@ -0,0 +1,143 @@
+            name = selectable.name
+        else:
+            name = self.name
+
+        co = ColumnClause(name, self.type)
+        co.key = self.name
+        co._proxies = [self]
+        if selectable._is_clone_of is not None:
+            co._is_clone_of = selectable._is_clone_of.columns.get(co.key)
+        co.table = selectable
+        co.named_with_table = False
+        if attach:
+            selectable._columns[co.key] = co
+        return co
+
+
+@compiles(FunctionColumn)
+def _compile_function_column(element, compiler, **kw):
+    if kw.get('asfrom', False):
+        return "(%s).%s" % (
+            compiler.process(element.function, **kw),
+            compiler.preparer.quote(element.name),
+        )
+    else:
+        return element.name
+
+
+class ColumnFunction(functions.FunctionElement):
+    __visit_name__ = 'function'
+
+    @property
+    def columns(self):
+        return FromClause.columns.fget(self)
+
+    def _populate_column_collection(self):
+        for name in self.column_names:
+            self._columns[name] = FunctionColumn(self, name)
+
+
+class jsonb_each_func(ColumnFunction):
+    name = 'jsonb_each'
+    column_names = ['key', 'value']
+
+
+@compiles(jsonb_each_func)
+def _compile_jsonb_each_func(element, compiler, **kw):
+    return compiler.visit_function(element, **kw)
+
+
+class AnnotationsInfoHandler(BaseHandler):
+    @auth_or_token
+    def get(self):
+        """
+        ---
+        description: Collects valid annotation origin/key pairs to filter on for scanning
+        responses:
+          200:
+            content:
+              application/json:
+                schema:
+                  allOf:
+                    - $ref: '#/components/schemas/Success'
+                    - type: object
+                      description: |
+                        An object in which each key is an annotation origin, and
+                        the values are arrays of { key: value_type } objects
+        """
+        user_accessible_group_ids = [g.id for g in self.current_user.accessible_groups]
+        user_accessible_filter_ids = [
+            filtr.id
+            for g in self.current_user.accessible_groups
+            for filtr in g.filters
+            if g.filters is not None
+        ]
+
+        annotations = jsonb_each_func(Annotation.data)
+        q = (
+            DBSession()
+            .query(Annotation.origin)
+            .add_columns(
+                annotations.c.key, func.jsonb_typeof(annotations.c.value).label("type")
+            )
+            .outerjoin(annotations, literal(True))
+            .join(Obj)
+            .filter(
+                Obj.id.in_(
+                    DBSession()
+                    .query(Candidate.obj_id)
+                    .filter(Candidate.filter_id.in_(user_accessible_filter_ids))
+                )
+            )
+            .join(GroupAnnotation)
+            .filter(GroupAnnotation.group_id.in_(user_accessible_group_ids))
+        )
```

Comment:
What exactly is happening here? Can you add an inline comment with example input and output ?

Reply:
I can't quite follow this all, either -- comments would be helpful